### PR TITLE
add resolveConfigWithFilePath API

### PR DESCRIPTION
**Pull request by SavePointSam**
This PR introduces a new API `prettier.resolveConfigWithFilePath`. This method uses the pre-existing internals to find and merge prettier configurations, but exposes the resolved file path for consumers. This will close https://github.com/prettier/prettier/issues/3451 and allow us to solve https://github.com/prettier/prettier-atom/issues/270.

I also reorganized the API docs a bit so that similar API sit next to each other.

### CODE REVIEW

**Comment by azz** [comment]
What if we just had `resolveConfig` return the path, too:

```js
prettier.resolveConfig.sync('some/path.js');
// -> 
//    {
//      useTabs: true,
//      configPath: "/abs/path/.prettierrc"
//    }
```

**Comment by SavePointSam** [comment]
That's also an option, however I would suggest against it. It makes sense that the API returns the configuration as Prettier can later ingest it. Adding extra properties in this one case doesn't quite make sense to me and feels like a mix of concerns. Even more so when the rest of the time we are working with the configuration payload it is separated between `config` and `filePath`.

File: src/config/resolve-config.js [hunk 1]
```diff
@@ -52,6 +62,13 @@ function _resolveConfig(filePath, opts, sync) {
       return null;
     }
 
+    if (!!includeFilePath) {
```
Comment (j-f1) [inline src/config/resolve-config.js]: Can this just be `if (includeFilePath)`?

Comment (SavePointSam) [reply src/config/resolve-config.js]: fixed!

File: docs/api.md [hunk 2]
```diff
@@ -57,6 +57,18 @@ If `options.editorconfig` is `true` and an [`.editorconfig` file](http://editorc
 
 Use `prettier.resolveConfig.sync(filePath [, options])` if you'd like to use sync version.
 
+## `prettier.resolveConfigWithFilePath(filePath [, options])`
+
+`resolveConfigWithFilePath` works that same as `resolveConfig` with one difference. The response will include both the configuration that was resolved, along with the file path to the configuration that was used to generate the configuration. However, it will not include the path to the located `.editorconfig` file when used as the core Prettier configuration acts as an override. This is useful for editor integrations, to subscribe to the resolved configuration and reload with any changes.
```
Comment (josephfrazier) [inline docs/api.md]: > The response will include both the configuration that was resolved, along with the file path to the configuration that was used to generate the configuration.

**Comment by josephfrazier** [comment]
...wait a minute. Couldn't we just export `resolveConfigFile` from `src/config/resolve-config.js` and `index.js`, instead of adding a separate function?

After typing this out, I realized that the use case (editor integrations) will likely want both the config and the file path anyway, so it probably makes sense to combine them into a single function, instead of forcing the integration to call two functions (which would also unnecessarily repeat the path resolution). I'm going to leave this comment here anyway, for the record.

**Comment by SavePointSam** [comment]
I thought the same thing at first. Though as someone who would be implementing this type of functionality, I'd much rather make a call that returns both the resolved/merged config as well as the file path of the config used. There is room for error when using those API separately.

**Comment by olsonpm** [comment]
I believe a valid workaround until this is resolved would be to use [cosmiconfig's `load`](https://github.com/prettier/prettier/pull/3846) directly.  Anyway, I'm planning on using that to address "[Show which options are used on format](https://github.com/prettier/prettier-atom/issues/178)".

**Comment by azz** [comment]
Is there a way to expose the path to the editorconfig file too?

**Comment by josephfrazier** [comment]
I don't think https://github.com/editorconfig/editorconfig-core-js provides a way to see the path of the files being parsed (note that there can be more than one). However, assuming there's just one, we could use some combination of `find-up` and `find-project-root` to locate it. See also https://github.com/prettier/prettier/pull/3559

**Comment by olsonpm** [comment]
"assuming just one" could produce false positives though the simplification would be nice.  Seems the devs behind editorconfig-core-js are pretty responsive though, so a PR adding a method such as this one wouldn't be out of the question.

**Comment by SavePointSam** [comment]
Sorry for the silence! Having some crunch time at the office :(

I totally agree we should find a way to expose the path to `.editorconfig`. Can it actually pick up multiple to merge together? That's pretty rough..

**Comment by olsonpm** [comment]
Bummer about the office time.  I'm in between jobs so I finally get to work on stuff I enjoy haha.

> Can it actually pick up multiple to merge together?

Unfortunately.  I created a [quick repo](https://github.com/olsonpm/test-editorconfig) if you wanted to mess around with it.

**Comment by duailibe** [comment]
FTR this was done in #4364, did not include the editorconfig part.

**Comment by suchipi** [comment]
I'm going to close this based on https://github.com/prettier/prettier/pull/3846#issuecomment-384495417

**Pull request closed without merging**